### ISSUE TITLE
torch multi-gpu training: port option missing

### ISSUE THREAD

**Issue by kuacakuaca** [rwth-i6/i6_core#459]
Currently it's not possible to run multi-gpu training jobs on the same host because of port conflict. It requires adding additional torchrun 
parameters.
https://github.com/rwth-i6/i6_core/blob/e5c9e241ef67e69098a18434bb1349f2db890939/returnn/training.py#L253C28-L253C28
https://pytorch.org/docs/stable/elastic/run.html#stacked-single-node-multi-worker

**Comment by albertz** [comment]
@Judyxujj how did you do that?

**Comment by albertz** [comment]
> Currently it's not possible to run multi-gpu training jobs on the same host because of port conflict.

Can you be more specific? What error do you get?

> It requires adding additional torchrun parameters.

Can you be more specific? What additional parameters? In [the doc you linked](https://pytorch.org/docs/stable/elastic/run.html#stacked-single-node-multi-worker), it just mentions `--standalone` for single-node multi-worker, nothing else?

**Comment by albertz** [comment]
I get such errors:
```[E ProcessGroupGloo.cpp:138] Gloo connectFullMesh failed with [../third_party/gloo/gloo/transport/tcp/pair.cc:534] Connection closed by peer [134.61.201.231]:927
[E ProcessGroupGloo.cpp:138] Gloo connectFullMesh failed with [../third_party/gloo/gloo/transport/tcp/pair.cc:534] Connection closed by peer [134.61.201.231]:18346
```
This happens via:
```
  File "/rwthfs/rz/cluster/home/az668407/setups/combined/2021-05-31/tools/returnn/returnn/datasets/basic.py", line 251, in _get_default_random_seed_offset
    return returnn.torch.distributed.get_ctx().rank() * 16127
```
Is this related to what you write here?

**Comment by albertz** [comment]
Ah, I think I misunderstood. Your issue is because multiple (independent) distributed jobs were scheduled on the same node, which caused the port conflict?

**Comment by albertz** [comment]
Now I got:
```
  File "/rwthfs/rz/cluster/home/az668407/setups/combined/2021-05-31/tools/returnn/returnn/torch/engine.py", line 198, in Engine.init_train_from
_config
    line: self._ddp_pt_model = self._torch_distributed_class(
              self._pt_model, device_ids=get_device_ids(), **self._torch_distributed_options
          )
  File "/rwthfs/rz/cluster/work/az668407/py-envs/py3.10-torch2.1/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 795, in DistributedDataParallel.__init__
    line: _verify_param_shape_across_processes(self.process_group, parameters)
  File "/rwthfs/rz/cluster/work/az668407/py-envs/py3.10-torch2.1/lib/python3.10/site-packages/torch/distributed/utils.py", line 265, in _verify_param_shape_across_processes 
    line: return dist._verify_params_across_processes(process_group, tensors, logger)
...
DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1333, unhandled system error (run with NCCL_DEBUG=INFO for details), NCCL version 2.18.1 
ncclSystemError: System call (e.g. socket, malloc) or external library call failed or device error. 
Last error:
socketStartConnect: Connect to fe80::ba59:9f03:fc:765c%ib0<57829> failed : Cannot assign requested address
```
Again not sure if this is related.

**Comment by kuacakuaca** [comment]
> What I read is that using `--standalone` should also automatically do that? (In case of single node only, though.)

yes, it works. Setting `--rdzv-backend=c10d` and `--rdzv-endpoint=localhost:$port_number` also works.

**Comment by albertz** [comment]
> Setting `--rdzv-backend=c10d` and `--rdzv-endpoint=localhost:$port_number` also works.

What would `$port_number` be? You cannot hardcode anything here, otherwise it would get conflicts again? Or would you just set `0` and it would automatically chose any free port?

**Comment by kuacakuaca** [comment]
> > Setting `--rdzv-backend=c10d` and `--rdzv-endpoint=localhost:$port_number` also works.
> 
> What would `$port_number` be? You cannot hardcode anything here, otherwise it would get conflicts again? Or would you just set `0` and it would automatically chose any free port?

yes, the documentation https://pytorch.org/docs/stable/elastic/run.html#stacked-single-node-multi-worker says:
To run multiple instances (separate jobs) of single-node, multi-worker on the same host, we need to make sure that each instance (job) is setup on different ports to avoid port conflicts (or worse, two jobs being merged as a single job). To do this you have to run with --rdzv-backend=c10d and specify a different port by setting --rdzv-endpoint=localhost:$PORT_k. For --nodes=1, its often convenient to let torchrun pick a free random port automatically instead of manually assigning different ports for each run.

```
torchrun
    --rdzv-backend=c10d
    --rdzv-endpoint=localhost:0
    --nnodes=1
    --nproc-per-node=$NUM_TRAINERS
    YOUR_TRAINING_SCRIPT.py (--arg1 ... train script args...)

```

**Comment by kuacakuaca** [comment]
setting `--standalone` should be easier, but i have no idea how it would effect the distributed training performance.

**Comment by albertz** [comment]
> setting `--standalone` should be easier, but i have no idea how it would effect the distributed training performance.

From the doc, I thought it would just do exactly the same?

**Comment by albertz** [comment]
So it sounds like `--standalone` is what we should use with `--nnodes=1`, right?

However, I wonder about what to do when using multiple nodes?
